File: mobile/src/services/llm.ts
/**
 * 大模型适配：与桌面版 carddeck/llm.py 同提示词、同 URL 规则、同解析逻辑。
 * 网络调用经注入的 HttpAdapter 发起（uni.request 适配 / 单测 stub）， Key 由调用方从安全存储传入。
 */
import type { ContactFields, LlmConfig, LlmProviderId } from '../types.js';
import { PROVIDERS } from '../types.js';

export const FIELDS = ['name', 'company', 'title', 'phone1', 'phone2', 'email', 'address',
  'business', 'event', 'met_at', 'notes', 'tags_printed', 'tags_inferred'] as const;

export type FieldKey = (typeof FIELDS)[number];

export const PROMPT = `你是名片信息提取器。看这张已裁切校正的单张名片图，只提取图上明确可见的信息。
规则：
1. 看不清或没有的信息留空字符串（标签留空数组），不要编造、不要补全号码。
2. 电话保留原始写法（含 +86、空格、连字符）；邮箱原样抄录。
3. tags_printed 只收录名片上明确印出的行业/业务词；模型推断的一律放 tags_inferred，不要混入事实字段。
4. business/event/met_at/notes 若图上没有就留空，等人工在表格里补充背景信息。
只返回 JSON，不要 markdown：
{"name":"","company":"","title":"","phone1":"","phone2":"","email":"","address":"","business":"","event":"","met_at":"","notes":"","tags_printed":[],"tags_inferred":[]}`;

export const MULTI_PROMPT = `这张照片里有多张名片（2～10 张）。请逐张提取每张名片上明确可见的信息，不要把 A 的电话填给 B。
规则：
1. 看不清或没有的信息留空字符串（标签留空数组），不要编造、不要补全号码。
2. 电话保留原始写法（含 +86、空格、连字符）；邮箱原样抄录。
3. tags_printed 只收录名片上明确印出的行业/业务词；模型推断的一律放 tags_inferred。
4. 按从左到右、从上到下的顺序排列。
只返回 JSON，不要 markdown：
{"cards": [{"name":"","company":"","title":"","phone1":"","phone2":"","email":"","address":"","business":"","event":"","met_at":"","notes":"","tags_printed":[],"tags_inferred":[]}]}`;

export function normalizeChatUrl(apiUrl: string): string {
  const u = (apiUrl ?? '').trim().replace(/\/+$/, '');
  if (u.endsWith('/chat/completions')) return u;
  if (u.endsWith('/completions')) return u;
  if (u.endsWith('/v1')) {
    return u.includes('deepseek') ? u.replace('/v1', '') + '/chat/completions' : u + '/chat/completions';
  }
  return u + '/chat/completions';
}

export function modelsUrl(apiUrl: string): string {
  const u = (apiUrl ?? '').trim().replace(/\/+$/, '');
  if (u.includes('/chat/completions')) return u.replace('/chat/completions', '/models');
  if (u.endsWith('/completions')) return u.slice(0, -'/completions'.length) + '/models';
  return u + '/models';
}

export function modelsCandidates(apiUrl: string): string[] {
  const first = modelsUrl(apiUrl);
  const cands = [first];
  if (first.includes('/v1/models')) {
    const alt = first.replace('/v1/models', '/models');
    if (!cands.includes(alt)) cands.push(alt);
  } else if (first.replace(/\/+$/, '').endsWith('/models')) {
    const alt = first.replace(/\/+$/, '').slice(0, -'/models'.length) + '/v1/models';
    if (!cands.includes(alt)) cands.push(alt);
  }
  return cands;
}

export function matchProvider(apiUrl: string): LlmProviderId {
  const u = (apiUrl ?? '').replace(/\/+$/, '');
  for (const k of ['deepseek', 'dashscope', 'openai', 'megmeet'] as const) {
    const p = PROVIDERS[k].replace(/\/+$/, '');
    if (u === p || u.startsWith(p.replace('/chat/completions', ''))) return k;
  }
  return 'custom';
}

export function emptyFields(mock = false): ContactFields & { _mock?: boolean } {
  const d = {} as ContactFields & { _mock?: boolean };
  for (const k of FIELDS) (d as unknown as Record<string, unknown>)[k] = k.startsWith('tags') ? [] : '';
  if (mock) d._mock = true;
  return d;
}

export function sanitizeFields(data: unknown): ContactFields {
  const out = emptyFields();
  if (typeof data !== 'object' || data === null) return out;
  const rec = data as Record<string, unknown>;
  for (const k of FIELDS) {
    if (k in rec) {
      const v = rec[k];
      (out as unknown as Record<string, unknown>)[k] = v == null ? (k.startsWith('tags') ? [] : '') : v;
    }
  }
  for (const k of ['tags_printed', 'tags_inferred'] as const) {
    const v = out[k] as unknown;
    if (typeof v === 'string') out[k] = v ? [v] : [];
    else if (!Array.isArray(v)) out[k] = [];
  }
  return out;
}

/** 去 markdown 围栏后取 JSON（与桌面版 _chat 后半段一致）。 */
export function parseChatContent(content: string): unknown {
  const t = content.trim().replace(/^```(?:json)?|```$/gm, '').trim();
  try {
    return JSON.parse(t);
  } catch {
    const m = /\{.*\}/s.exec(t);
    return m ? JSON.parse(m[0]) : {};
  }
}

export interface ChatMessage {
  role: 'user';
  content: Array<{ type: 'text'; text: string } | { type: 'image_url'; image_url: { url: string } }>;
}

export function buildChatPayload(model: string, text: string, dataUrl: string): object {
  const content: ChatMessage['content'] = [
    { type: 'text', text },
    { type: 'image_url', image_url: { url: dataUrl } },
  ];
  const messages: ChatMessage[] = [{ role: 'user', content }];
  return {
    model,
    temperature: 0,
    response_format: { type: 'json_object' },
    messages,
  };
}

export function extractCards(data: unknown): ContactFields[] {
  let cards: unknown = [];
  if (Array.isArray(data)) cards = data;
  else if (typeof data === 'object' && data !== null) cards = (data as Record<string, unknown>).cards ?? [];
  if (!Array.isArray(cards)) return [];
  return (cards as unknown[]).slice(0, 10).map(sanitizeFields);
}

export function httpErrorDetail(status: number | null, body: string, errText: string): string {
  if (status == null) return errText.slice(0, 300);
  let hint = '';
  if (status === 401) hint = '（Key 未通过：检查是否复制完整、有无多余空格、是否过期、是否有该模型调用权限）';
  else if (status === 403) hint = '（无权限：Key 可能无此模型/地域权限）';
  else if (status === 404) hint = '（地址可能不对：检查 URL 路径）';
  return `HTTP ${status} ${(body ?? '').slice(0, 300)}${hint}`;
}

export interface HttpAdapter {
  postJson(url: string, apiKey: string, payload: object, timeoutMs: number): Promise<{ status: number; bodyText: string }>;
  get(url: string, apiKey: string, timeoutMs: number): Promise<{ status: number; bodyText: string }>;
}

function choicesText(bodyText: string): string {
  try {
    return (JSON.parse(bodyText) as { choices: Array<{ message: { content: string } }> }).choices[0].message.content;
  } catch {
    return bodyText.slice(0, 200);
  }
}

export async function recognizePhotoMulti(
  http: HttpAdapter, imageDataUrl: string, cfg: LlmConfig, timeoutMs = 120000,
): Promise<ContactFields[]> {
  const url = normalizeChatUrl(cfg.api_url);
  if (!url || !cfg.api_key) return [];
  const r = await http.postJson(url, cfg.api_key, buildChatPayload(cfg.model, MULTI_PROMPT, imageDataUrl), timeoutMs);
  if (r.status < 200 || r.status >= 300) throw new Error(httpErrorDetail(r.status, r.bodyText, '识别失败'));
  return extractCards(parseChatContent(choicesText(r.bodyText)));
}

export async function recognizeCrop(
  http: HttpAdapter, imageDataUrl: string, cfg: LlmConfig, timeoutMs = 90000,
): Promise<ContactFields & { _mock?: boolean; _error?: string }> {
  const url = normalizeChatUrl(cfg.api_url);
  if (!url || !cfg.api_key) return emptyFields(true);
  const r = await http.postJson(url, cfg.api_key, buildChatPayload(cfg.model, PROMPT, imageDataUrl), timeoutMs);
  if (r.status < 200 || r.status >= 300) {
    const e = emptyFields();
    (e as ContactFields & { _error?: string })._error = `识别失败：${httpErrorDetail(r.status, r.bodyText, '')}`;
    return e;
  }
  return sanitizeFields(parseChatContent(choicesText(r.bodyText)));
}

export async function listModels(
  http: HttpAdapter, apiUrl: string, apiKey: string, timeoutMs = 30000,
): Promise<{ models: string[]; url: string }> {
  const errs: string[] = [];
  for (const url of modelsCandidates(normalizeChatUrl(apiUrl))) {
    const r = await http.get(url, apiKey, timeoutMs);
    if (r.status < 200 || r.status >= 300) {
      errs.push(`${url}：${httpErrorDetail(r.status, r.bodyText, '')}`);
      continue;
    }
    let items: unknown[] = [];
    try {
      const d = JSON.parse(r.bodyText) as { data?: unknown };
      if (d && Array.isArray(d.data)) items = d.data;
    } catch { /* 空列表 */ }
    const ids = [...new Set(items
      .filter((x): x is Record<string, unknown> => typeof x === 'object' && x !== null && 'id' in x)
      .map((x) => String(x.id)))].sort();
    return { models: ids, url };
  }
  throw new Error(errs.join('；').slice(0, 500) || '网关未提供模型列表接口，请手填模型名');
}

export async function testConnection(
  http: HttpAdapter, apiUrl: string, apiKey: string, model: string, timeoutMs = 30000,
): Promise<{ ok: true; reply: string }> {
  const url = normalizeChatUrl(apiUrl);
  const r = await http.postJson(url, apiKey, {
    model, temperature: 0, max_tokens: 16, messages: [{ role: 'user', content: '只回复 ok' }],
  }, timeoutMs);
  if (r.status < 200 || r.status >= 300) throw new Error(httpErrorDetail(r.status, r.bodyText, '连接失败'));
  return { ok: true, reply: choicesText(r.bodyText).slice(0, 200) };
}

File: carddeck/llm.py
"""视觉 LLM 适配器：OpenAI 兼容 /chat/completions，图片 base64 输入，结构化 JSON 输出。"""
import base64
import json
import os
import re

import requests

from .llm_config import get_llm_config

FIELDS = ["name", "company", "title", "phone1", "phone2", "email", "address",
          "business", "event", "met_at", "notes", "tags_printed", "tags_inferred"]

PROMPT = """你是名片信息提取器。看这张已裁切校正的单张名片图，只提取图上明确可见的信息。
规则：
1. 看不清或没有的信息留空字符串（标签留空数组），不要编造、不要补全号码。
2. 电话保留原始写法（含 +86、空格、连字符）；邮箱原样抄录。
3. tags_printed 只收录名片上明确印出的行业/业务词；模型推断的一律放 tags_inferred，不要混入事实字段。
4. business/event/met_at/notes 若图上没有就留空，等人工在表格里补充背景信息。
只返回 JSON，不要 markdown：
{"name":"","company":"","title":"","phone1":"","phone2":"","email":"","address":"","business":"","event":"","met_at":"","notes":"","tags_printed":[],"tags_inferred":[]}"""

MULTI_PROMPT = """这张照片里有多张名片（2～10 张）。请逐张提取每张名片上明确可见的信息，不要把 A 的电话填给 B。
规则：
1. 看不清或没有的信息留空字符串（标签留空数组），不要编造、不要补全号码。
2. 电话保留原始写法（含 +86、空格、连字符）；邮箱原样抄录。
3. tags_printed 只收录名片上明确印出的行业/业务词；模型推断的一律放 tags_inferred。
4. 按从左到右、从上到下的顺序排列。
只返回 JSON，不要 markdown：
{"cards": [{"name":"","company":"","title":"","phone1":"","phone2":"","email":"","address":"","business":"","event":"","met_at":"","notes":"","tags_printed":[],"tags_inferred":[]}]}"""


def _sanitize_fields(data: dict) -> dict:
    out = _empty_result()
    if not isinstance(data, dict):
        return out
    for k in FIELDS:
        if k in data:
            out[k] = data[k] if data[k] is not None else ([] if k.startswith("tags") else "")
    for k in ("tags_printed", "tags_inferred"):
        if isinstance(out[k], str):
            out[k] = [out[k]] if out[k] else []
        elif not isinstance(out[k], list):
            out[k] = []
    return out


def _chat(api_url, api_key, model, text, image_path, timeout):
    data_url = _img_to_data_url(image_path)
    r = requests.post(api_url, headers={"Authorization": f"Bearer {api_key}", "Content-Type": "application/json"},
                      json={"model": model, "temperature": 0, "response_format": {"type": "json_object"},
                            "messages": [{"role": "user", "content": [
                                {"type": "text", "text": text},
                                {"type": "image_url", "image_url": {"url": data_url}}]}]},
                      timeout=timeout)
    try:
        r.raise_for_status()
    except Exception as e:
        raise RuntimeError(_http_error_detail(e))
    content = r.json()["choices"][0]["message"]["content"]
    content = re.sub(r"^```(?:json)?|```$", "", content.strip(), flags=re.MULTILINE).strip()
    try:
        return json.loads(content)
    except Exception:
        m = re.search(r"\{.*\}", content, re.DOTALL)
        return json.loads(m.group(0)) if m else {}


def recognize_photo_multi(photo_path: str, timeout=120, api_url="", api_key="", model="") -> list:
    """免框选：整图直接识别多张名片，返回 [fields...]。未配置 Key 时返回 []。"""
    if not api_url or not api_key:
        cfg = get_llm_config()
        api_url = api_url or cfg["api_url"]
        api_key = api_key or cfg["api_key"]
        model = model or cfg["model"]
    api_url = _normalize_chat_url(api_url)
    if not api_url or not api_key:
        return []
    data = _chat(api_url, api_key, model, MULTI_PROMPT, photo_path, timeout)
    cards = data.get("cards", []) if isinstance(data, dict) else []
    if isinstance(data, list):
        cards = data
    return [_sanitize_fields(c) for c in cards[:10]] if isinstance(cards, list) else []


def _img_to_data_url(path: str) -> str:
    from PIL import Image
    import io
    im = Image.open(path).convert("RGB")
    im.thumbnail((1400, 1400))
    buf = io.BytesIO()
    im.save(buf, format="JPEG", quality=88)
    b64 = base64.b64encode(buf.getvalue()).decode()
    return "data:image/jpeg;base64," + b64


def _empty_result(mock=False):
    d = {k: ([] if k.startswith("tags") else "") for k in FIELDS}
    if mock:
        d["_mock"] = True
    return d


def recognize_crop(crop_path: str, timeout=90, api_url="", api_key="", model="") -> dict:
    """逐张调用视觉模型。未配置 API Key 时返回空模板（标记 _mock），保证离线可走通全流程。"""
    if not api_url or not api_key:
        cfg = get_llm_config()
        api_url = api_url or cfg["api_url"]
        api_key = api_key or cfg["api_key"]
        model = model or cfg["model"]
    api_url = _normalize_chat_url(api_url)
    if not api_url or not api_key:
        return _empty_result(mock=True)
    # 注意：DeepSeek 官方只有 deepseek-v4-flash-vision-exp 支持图片；
    # deepseek-chat/reasoner、v4-flash/v4-pro 仍是纯文本，直接调会 400。
    # 这里不硬拦：统一真实调用，失败信息原样返回，方便按网关实际能力判断。
    data_url = _img_to_data_url(crop_path)
    headers = {"Authorization": f"Bearer {api_key}", "Content-Type": "application/json"}
    payload = {
        "model": model,
        "temperature": 0,
        "response_format": {"type": "json_object"},
        "messages": [{"role": "user", "content": [
            {"type": "text", "text": PROMPT},
            {"type": "image_url", "image_url": {"url": data_url}},
        ]}],
    }
    r = requests.post(api_url, headers=headers, json=payload, timeout=timeout)
    try:
        r.raise_for_status()
    except Exception as e:
        raise RuntimeError(_http_error_detail(e))
    content = r.json()["choices"][0]["message"]["content"]
    content = re.sub(r"^```(?:json)?|```$", "", content.strip(), flags=re.MULTILINE).strip()
    try:
        data = json.loads(content)
    except Exception:
        m = re.search(r"\{.*\}", content, re.DOTALL)
        data = json.loads(m.group(0)) if m else {}
    out = _empty_result()
    for k in FIELDS:
        if k in data:
            out[k] = data[k] if data[k] is not None else ([] if k.startswith("tags") else "")
    for k in ("tags_printed", "tags_inferred"):
        if isinstance(out[k], str):
            out[k] = [out[k]] if out[k] else []
    return out


def _normalize_chat_url(api_url: str) -> str:
    u = (api_url or "").strip().rstrip("/")
    if u.endswith("/chat/completions"):
        return u
    if u.endswith("/completions"):
        return u
    # 只填了域名/base（如 https://api.deepseek.com 或 .../v1）时自动补全
    if u.endswith("/v1"):
        # DeepSeek 的 base 就是 https://api.deepseek.com（/v1 仅兼容用），补 /chat/completions
        return u.replace("/v1", "") + "/chat/completions" if "deepseek" in u else u + "/chat/completions"
    return u + "/chat/completions"


def _models_url(api_url: str) -> str:
    u = (api_url or "").strip().rstrip("/")
    if "/chat/completions" in u:
        return u.replace("/chat/completions", "/models")
    if u.endswith("/completions"):
        return u[: -len("/completions")] + "/models"
    return u + "/models"


def _http_error_detail(e: Exception) -> str:
    resp = getattr(e, "response", None)
    if resp is not None:
        try:
            body = resp.text or ""
        except Exception:
            body = ""
        code = getattr(resp, "status_code", "?")
        hint = ""
        if code == 401:
            hint = "（Key 未通过：检查是否复制完整、有无多余空格、是否过期、是否有该模型调用权限）"
        elif code == 403:
            hint = "（无权限：Key 可能无此模型/地域权限）"
        elif code == 404:
            hint = "（地址可能不对：检查 URL 中的 compatible-mode/v1 路径）"
        return f"HTTP {code} {body[:300]}{hint}"
    return str(e)[:300]


def list_models(api_url: str, api_key: str, timeout=30) -> dict:
    """OpenAI 兼容 GET /models。返回 {"models": [id...]}。"""
    import requests as _rq
    url = _models_url(_normalize_chat_url(api_url))
    try:
        r = _rq.get(url, headers={"Authorization": f"Bearer {api_key}"}, timeout=timeout)
        r.raise_for_status()
    except Exception as e:
        raise RuntimeError(_http_error_detail(e))
    data = r.json()
    items = data.get("data", []) if isinstance(data, dict) else []
    ids = sorted({str(x.get("id")) for x in items if isinstance(x, dict) and x.get("id")})
    return {"models": ids, "url": url}


def test_connection(api_url: str, api_key: str, model: str, timeout=30) -> dict:
    """只发一句纯文本测试连通性，不传图片，省 token。"""
    import requests as _rq
    api_url = _normalize_chat_url(api_url)
    try:
        r = _rq.post(api_url, headers={"Authorization": f"Bearer {api_key}", "Content-Type": "application/json"},
                     json={"model": model, "temperature": 0, "max_tokens": 16,
                           "messages": [{"role": "user", "content": "只回复 ok"}]},
                     timeout=timeout)
        r.raise_for_status()
    except Exception as e:
        raise RuntimeError(_http_error_detail(e))
    try:
        txt = r.json()["choices"][0]["message"]["content"]
    except Exception:
        txt = r.text[:200]
    return {"ok": True, "reply": (txt or "")[:200]}
